diff --git a/src/data/ax/chansey.h b/src/data/ax/chansey.h
index eb2f96203..7177eb409 100644
--- a/src/data/ax/chansey.h
+++ b/src/data/ax/chansey.h
@@ -1,3 +1,4 @@
+#include "shared_anims.h"
 static const axmain sAxMainChansey;
 const SiroArchive gAxChansey = {"SIRO", &sAxMainChansey};
 
@@ -1662,25 +1663,7 @@ static const ax_anim sChanseyAnims_9_8[] = {
 	{ .frames = 1, .unkFlags = 0, .poseId = 171, .offset = {0, 0}, .shadow = {0, 0} },
 	AX_ANIM_TERMINATOR
 };
-static const ax_anim sChanseyAnims_10_1[] = {
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {13, 0}, .shadow = {13, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-13, 0}, .shadow = {-13, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	AX_ANIM_TERMINATOR
-};
+#define sChanseyAnims_10_1 sSharedAnims_10_SwayPose178
 static const ax_anim sChanseyAnims_10_2[] = {
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {0, 0}, .shadow = {0, 0} },
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {6, -6}, .shadow = {6, -6} },
diff --git a/src/data/ax/electabuzz.h b/src/data/ax/electabuzz.h
index 48b26cbfd..ad805d161 100644
--- a/src/data/ax/electabuzz.h
+++ b/src/data/ax/electabuzz.h
@@ -1,3 +1,4 @@
+#include "shared_anims.h"
 static const axmain sAxMainElectabuzz;
 const SiroArchive gAxElectabuzz = {"SIRO", &sAxMainElectabuzz};
 
@@ -1638,25 +1639,7 @@ static const ax_anim sElectabuzzAnims_9_8[] = {
 	{ .frames = 1, .unkFlags = 0, .poseId = 171, .offset = {0, 0}, .shadow = {0, 0} },
 	AX_ANIM_TERMINATOR
 };
-static const ax_anim sElectabuzzAnims_10_1[] = {
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {13, 0}, .shadow = {13, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-13, 0}, .shadow = {-13, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	AX_ANIM_TERMINATOR
-};
+#define sElectabuzzAnims_10_1 sSharedAnims_10_SwayPose178
 static const ax_anim sElectabuzzAnims_10_2[] = {
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {0, 0}, .shadow = {0, 0} },
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {6, -6}, .shadow = {6, -6} },
diff --git a/src/data/ax/porygon.h b/src/data/ax/porygon.h
index ccd3c0770..c5be5dd76 100644
--- a/src/data/ax/porygon.h
+++ b/src/data/ax/porygon.h
@@ -1,3 +1,4 @@
+#include "shared_anims.h"
 static const axmain sAxMainPorygon;
 const SiroArchive gAxPorygon = {"SIRO", &sAxMainPorygon};
 
@@ -1602,25 +1603,7 @@ static const ax_anim sPorygonAnims_9_8[] = {
 	{ .frames = 1, .unkFlags = 0, .poseId = 171, .offset = {0, 0}, .shadow = {0, 0} },
 	AX_ANIM_TERMINATOR
 };
-static const ax_anim sPorygonAnims_10_1[] = {
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {13, 0}, .shadow = {13, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-13, 0}, .shadow = {-13, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	AX_ANIM_TERMINATOR
-};
+#define sPorygonAnims_10_1 sSharedAnims_10_SwayPose178
 static const ax_anim sPorygonAnims_10_2[] = {
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {0, 0}, .shadow = {0, 0} },
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {6, -6}, .shadow = {6, -6} },
diff --git a/src/data/ax/zigzagoon.h b/src/data/ax/zigzagoon.h
index 586ed7ddd..d03edf376 100644
--- a/src/data/ax/zigzagoon.h
+++ b/src/data/ax/zigzagoon.h
@@ -1,3 +1,4 @@
+#include "shared_anims.h"
 static const axmain sAxMainZigzagoon;
 const SiroArchive gAxZigzagoon = {"SIRO", &sAxMainZigzagoon};
 
@@ -1670,25 +1671,7 @@ static const ax_anim sZigzagoonAnims_9_8[] = {
 	{ .frames = 1, .unkFlags = 0, .poseId = 171, .offset = {0, 0}, .shadow = {0, 0} },
 	AX_ANIM_TERMINATOR
 };
-static const ax_anim sZigzagoonAnims_10_1[] = {
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {13, 0}, .shadow = {13, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-13, 0}, .shadow = {-13, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {12, 0}, .shadow = {12, 0} },
-	{ .frames = 3, .unkFlags = 0, .poseId = 178, .offset = {-12, 0}, .shadow = {-12, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {10, 0}, .shadow = {10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-10, 0}, .shadow = {-10, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {6, 0}, .shadow = {6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {-6, 0}, .shadow = {-6, 0} },
-	{ .frames = 2, .unkFlags = 0, .poseId = 178, .offset = {0, 0}, .shadow = {0, 0} },
-	AX_ANIM_TERMINATOR
-};
+#define sZigzagoonAnims_10_1 sSharedAnims_10_SwayPose178
 static const ax_anim sZigzagoonAnims_10_2[] = {
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {0, 0}, .shadow = {0, 0} },
 	{ .frames = 2, .unkFlags = 0, .poseId = 179, .offset = {6, -6}, .shadow = {6, -6} },
diff --git a/tools/aif2pcm/Makefile b/tools/aif2pcm/Makefile
index dfa04e4a8..96311d4be 100644
--- a/tools/aif2pcm/Makefile
+++ b/tools/aif2pcm/Makefile
@@ -1,24 +1,17 @@
-CC := /c/msys64/mingw64/bin/gcc
-
-CFLAGS = -Wall -Wextra -Wno-switch -Werror -std=c11 -O2
-
-LIBS = -lm
-
-SRCS = main.c extended.c
-
-ifeq ($(OS),Windows_NT)
-EXE := .exe
-else
-EXE :=
-endif
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
+CFLAGS := -Wall -Wextra -Wno-switch -Werror -std=c11 -O2 -pipe
+LIBS := -lm
 
 .PHONY: all clean
 
-all: aif2pcm$(EXE)
-	@:
-
-aif2pcm$(EXE): $(SRCS)
-	$(CC) $(CFLAGS) $(SRCS) -o $@ $(LDFLAGS) $(LIBS)
+all: aif2pcm
 
 clean:
-	$(RM) aif2pcm aif2pcm.exe
+	rm -f aif2pcm aif2pcm.exe *.o *.i *.s
+
+aif2pcm: main.o extended.o
+	$(CC) $(CFLAGS) -o $@ $^ $(LIBS)
+
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
diff --git a/tools/bin2c/Makefile b/tools/bin2c/Makefile
index 310ff9e43..f74d3e628 100644
--- a/tools/bin2c/Makefile
+++ b/tools/bin2c/Makefile
@@ -1,22 +1,17 @@
-CC := /c/msys64/mingw64/bin/gcc
-
-CFLAGS = -Wall -Wextra -Werror -std=c11 -O2
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
+CFLAGS := -Wall -Wextra -Werror -std=c11 -O2 -pipe
+LIBS := 
 
 .PHONY: all clean
 
-SRCS = bin2c.c
-
-ifeq ($(OS),Windows_NT)
-EXE := .exe
-else
-EXE :=
-endif
+all: bin2c
 
-all: bin2c$(EXE)
-	@:
+clean:
+	rm -f bin2c bin2c.exe *.o *.i *.s
 
-bin2c$(EXE): $(SRCS)
-	$(CC) $(CFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+bin2c: bin2c.o
+	$(CC) $(CFLAGS) -o $@ $^ $(LIBS)
 
-clean:
-	$(RM) bin2c bin2c.exe
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
diff --git a/tools/br_ips/Makefile b/tools/br_ips/Makefile
index 03b3b59e1..9f9c27f92 100644
--- a/tools/br_ips/Makefile
+++ b/tools/br_ips/Makefile
@@ -1,13 +1,20 @@
-CC := /c/msys64/mingw64/bin/gcc
-CFLAGS := -O3
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
+CFLAGS := -O3 -pipe
+LIBS := 
+
+.PHONY: all clean
 
 all: br_ips ips_patch
 
 clean:
-	rm -f br_ips ips_patch br_ips.exe ips_patch.exe
+	rm -f br_ips br_ips.exe ips_patch ips_patch.exe *.o *.i *.s
+
+br_ips: br_ips.o
+	$(CC) $(CFLAGS) -o $@ $^ $(LIBS)
 
-br_ips: br_ips.c
-	$(CC) $(CFLAGS) -o $@ $^
+ips_patch: ips_patch.o
+	$(CC) $(CFLAGS) -o $@ $^ $(LIBS)
 
-ips_patch: ips_patch.c
-	$(CC) $(CFLAGS) -o $@ $^
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
diff --git a/tools/dungeonjson/Makefile b/tools/dungeonjson/Makefile
index 561ba89fe..96bd80ee9 100644
--- a/tools/dungeonjson/Makefile
+++ b/tools/dungeonjson/Makefile
@@ -1,8 +1,10 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -Wall -std=c++11 -O2
+CXXFLAGS := -Wall -std=c++11 -O2 -pipe
 
 SRCS := json11.cpp dungeonjson.cpp move_id_json.cpp item_id_json.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := dungeonjson.h move_id_json.h
 
@@ -11,8 +13,11 @@ HEADERS := dungeonjson.h move_id_json.h
 all: dungeonjson
 	@:
 
-dungeonjson: $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+dungeonjson: $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) dungeonjson dungeonjson.exe
+	$(RM) dungeonjson dungeonjson.exe $(OBJS) *.i *.s
diff --git a/tools/gbafix/Makefile b/tools/gbafix/Makefile
index 1d5ef71b4..90dea55da 100644
--- a/tools/gbafix/Makefile
+++ b/tools/gbafix/Makefile
@@ -1,19 +1,17 @@
-CC := /c/msys64/mingw64/bin/gcc
-.PHONY: all clean
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
+CFLAGS := -Wall -Wextra -Werror -Wno-error=stringop-truncation -std=c11 -O2 -pipe
+LIBS := 
 
-SRCS = gbafix.c
+.PHONY: all clean
 
-ifeq ($(OS),Windows_NT)
-EXE := .exe
-else
-EXE :=
-endif
+all: gbafix
 
-all: gbafix$(EXE)
-	@:
+clean:
+	rm -f gbafix gbafix.exe *.o *.i *.s
 
-gbafix$(EXE): $(SRCS)
-	$(CC) $(SRCS) -o $@ $(LDFLAGS)
+gbafix: gbafix.o
+	$(CC) $(CFLAGS) -o $@ $^ $(LIBS)
 
-clean:
-	$(RM) gbafix gbafix.exe
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
diff --git a/tools/gbagfx/Makefile b/tools/gbagfx/Makefile
index 36b166c8a..caa0410d2 100644
--- a/tools/gbagfx/Makefile
+++ b/tools/gbagfx/Makefile
@@ -1,12 +1,13 @@
-CC := /c/msys64/mingw64/bin/gcc
-
-CFLAGS = -Wall -Wextra -Werror -Wno-sign-compare -std=c11 -DPNG_SKIP_SETJMP_CHECK
-CFLAGS += $(shell pkg-config --cflags libpng)
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
 
+CFLAGS = -Wall -Wextra -Werror -Wno-sign-compare -std=c11 -DPNG_SKIP_SETJMP_CHECK -O2 -pipe
 LIBS = -lpng -lz
-LDFLAGS += $(shell pkg-config --libs-only-L libpng)
 
 SRCS = main.c convert_png.c gfx.c jasc_pal.c lz.c rl.c util.c font.c huff.c
+OBJS = $(SRCS:.c=.o)
+
+.PHONY: all clean
 
 ifeq ($(OS),Windows_NT)
 EXE := .exe
@@ -14,18 +15,14 @@ else
 EXE :=
 endif
 
-.PHONY: all clean
-
 all: gbagfx$(EXE)
 	@:
 
-gbagfx-debug$(EXE): CFLAGS += -Og -g -DDEBUG
-gbagfx-debug$(EXE): $(SRCS) convert_png.h gfx.h global.h jasc_pal.h lz.h rl.h util.h font.h
-	$(CC) $(CFLAGS) $(SRCS) -o $@ $(LDFLAGS) $(LIBS)
+gbagfx$(EXE): $(OBJS)
+	$(CC) $(CFLAGS) $(OBJS) -o $@ $(LIBS)
 
-gbagfx$(EXE): CFLAGS += -O2
-gbagfx$(EXE): $(SRCS) convert_png.h gfx.h global.h jasc_pal.h lz.h rl.h util.h font.h
-	$(CC) $(CFLAGS) $(SRCS) -o $@ $(LDFLAGS) $(LIBS)
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
 
 clean:
-	$(RM) gbagfx gbagfx.exe gbagfx-debug gbagfx-debug.exe
+	$(RM) gbagfx gbagfx.exe $(OBJS) *.i *.s
diff --git a/tools/jsonproc/Makefile b/tools/jsonproc/Makefile
index 70ea76ca4..2ebcc0bb6 100755
--- a/tools/jsonproc/Makefile
+++ b/tools/jsonproc/Makefile
@@ -1,10 +1,12 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -Wall -std=c++11 -O2
+CXXFLAGS := -Wall -std=c++11 -O2 -pipe -I.
 
 INCLUDES := -I .
 
 SRCS := jsonproc.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := jsonproc.h inja.hpp nlohmann/json.hpp
 
@@ -19,8 +21,11 @@ endif
 all: jsonproc$(EXE)
 	@:
 
-jsonproc$(EXE): $(SRCS) $(HEADERS)
+jsonproc$(EXE): $(OBJS) $(HEADERS)
 	$(CXX) $(CXXFLAGS) $(INCLUDES) $(SRCS) -o $@ $(LDFLAGS)
 
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
+
 clean:
-	$(RM) jsonproc jsonproc.exe
+	$(RM) jsonproc jsonproc.exe $(OBJS) *.i *.s
diff --git a/tools/mapjson/Makefile b/tools/mapjson/Makefile
index 0df046ae3..80d545721 100644
--- a/tools/mapjson/Makefile
+++ b/tools/mapjson/Makefile
@@ -1,8 +1,10 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -Wall -std=c++11 -O2
+CXXFLAGS := -Wall -std=c++11 -O2 -pipe
 
 SRCS := json11.cpp mapjson.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := mapjson.h
 
@@ -17,8 +19,11 @@ endif
 all: mapjson$(EXE)
 	@:
 
-mapjson$(EXE): $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+mapjson$(EXE): $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) mapjson mapjson.exe
+	$(RM) mapjson mapjson.exe $(OBJS) *.i *.s
diff --git a/tools/mid2agb/Makefile b/tools/mid2agb/Makefile
index e6eb4ac91..a1151c76e 100644
--- a/tools/mid2agb/Makefile
+++ b/tools/mid2agb/Makefile
@@ -1,8 +1,10 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror
+CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror -pipe
 
 SRCS := agb.cpp error.cpp main.cpp midi.cpp tables.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := agb.h error.h main.h midi.h tables.h
 
@@ -17,8 +19,11 @@ endif
 all: mid2agb$(EXE)
 	@:
 
-mid2agb$(EXE): $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+mid2agb$(EXE): $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) mid2agb mid2agb.exe
+	$(RM) mid2agb mid2agb.exe $(OBJS) *.i *.s
diff --git a/tools/preproc/Makefile b/tools/preproc/Makefile
index 0e392e76a..4b039ba26 100644
--- a/tools/preproc/Makefile
+++ b/tools/preproc/Makefile
@@ -1,9 +1,11 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror
+CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror -pipe
 
 SRCS := asm_file.cpp c_file.cpp charmap.cpp preproc.cpp string_parser.cpp \
 	utf8.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := asm_file.h c_file.h char_util.h charmap.h preproc.h string_parser.h \
 	utf8.h
@@ -19,8 +21,11 @@ endif
 all: preproc$(EXE)
 	@:
 
-preproc$(EXE): $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+preproc$(EXE): $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) preproc preproc.exe
+	$(RM) preproc preproc.exe $(OBJS) *.i *.s
diff --git a/tools/ramscrgen/Makefile b/tools/ramscrgen/Makefile
index f386b0bed..634f0d2a6 100644
--- a/tools/ramscrgen/Makefile
+++ b/tools/ramscrgen/Makefile
@@ -1,8 +1,10 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror
+CXXFLAGS := -std=c++11 -O2 -Wall -Wno-switch -Werror -pipe
 
 SRCS := main.cpp sym_file.cpp elf.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := ramscrgen.h sym_file.h elf.h char_util.h
 
@@ -17,8 +19,11 @@ endif
 all: ramscrgen$(EXE)
 	@:
 
-ramscrgen$(EXE): $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+ramscrgen$(EXE): $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) ramscrgen ramscrgen.exe
+	$(RM) ramscrgen ramscrgen.exe $(OBJS) *.i *.s
diff --git a/tools/rsfont/Makefile b/tools/rsfont/Makefile
index eda38eb09..a930a1648 100644
--- a/tools/rsfont/Makefile
+++ b/tools/rsfont/Makefile
@@ -1,12 +1,14 @@
-CC := /c/msys64/mingw64/bin/gcc
+HOSTCC ?= /c/msys64/mingw64/bin/gcc
+CC := $(HOSTCC)
 
-CFLAGS = -Wall -Wextra -Werror -std=c11 -O2 -DPNG_SKIP_SETJMP_CHECK
+CFLAGS = -Wall -Wextra -Werror -std=c11 -O2 -pipe -DPNG_SKIP_SETJMP_CHECK
 CFLAGS += $(shell pkg-config --cflags libpng)
-
 LIBS = -lpng -lz
 LDFLAGS += $(shell pkg-config --libs-only-L libpng)
 
 SRCS = main.c convert_png.c util.c font.c
+OBJS = $(SRCS:.c=.o)
+HEADERS = convert_png.h gfx.h global.h util.h font.h
 
 .PHONY: all clean
 
@@ -19,8 +21,11 @@ endif
 all: rsfont$(EXE)
 	@:
 
-rsfont$(EXE): $(SRCS) convert_png.h gfx.h global.h util.h font.h
-	$(CC) $(CFLAGS) $(SRCS) -o $@ $(LDFLAGS) $(LIBS)
+rsfont$(EXE): $(OBJS) $(HEADERS)
+	$(CC) $(CFLAGS) $(OBJS) -o $@ $(LDFLAGS) $(LIBS)
+
+%.o: %.c
+	$(CC) $(CFLAGS) -c $< -o $@
 
 clean:
-	$(RM) rsfont rsfont.exe
+	$(RM) rsfont rsfont.exe $(OBJS) *.i *.s
diff --git a/tools/scaninc/Makefile b/tools/scaninc/Makefile
index 28635e2a1..d916514e8 100644
--- a/tools/scaninc/Makefile
+++ b/tools/scaninc/Makefile
@@ -1,8 +1,10 @@
-CXX := /c/msys64/mingw64/bin/g++
+HOSTCXX ?= /c/msys64/mingw64/bin/g++
+CXX := $(HOSTCXX)
 
-CXXFLAGS = -Wall -Werror -std=c++11 -O2
+CXXFLAGS = -Wall -Werror -std=c++11 -O2 -pipe
 
 SRCS = scaninc.cpp c_file.cpp asm_file.cpp source_file.cpp
+OBJS := $(SRCS:.cpp=.o)
 
 HEADERS := scaninc.h asm_file.h c_file.h source_file.h
 
@@ -17,8 +19,11 @@ endif
 all: scaninc$(EXE)
 	@:
 
-scaninc$(EXE): $(SRCS) $(HEADERS)
-	$(CXX) $(CXXFLAGS) $(SRCS) -o $@ $(LDFLAGS)
+scaninc$(EXE): $(OBJS) $(HEADERS)
+	$(CXX) $(CXXFLAGS) $(OBJS) -o $@ $(LDFLAGS)
+
+%.o: %.cpp
+	$(CXX) $(CXXFLAGS) -c $< -o $@
 
 clean:
-	$(RM) scaninc scaninc.exe
+	$(RM) scaninc scaninc.exe $(OBJS) *.i *.s
